Start AUS2200 documentation (#32)
coecms/cms-wiki: Wiki/models/models-aus2200.md

<script src="https://cdn.jsdelivr.net/gh/atteggiani/animated-terminal/animated-terminal.min.js" defer></script>

# AUS2200 - A complete guide

## Background

The AUS2200 model is a 0.0198 degree (2.2km at the equator) Unified Model vn12.2 configuration

This page assumes a familiarity with the general structure of rose-cylc suites and is not intended as a rose-cylc tutorial. 

## Setup
 * MOSRS access
 * Projects to join
    * rt52 - ERA5 data
    * zz93 - ERA5-Land data
    * hh5 - analysis software environment
    * access - `$UMDIR` and um2netcdf4 library
 * Ancil suite
 * Resource Requirements
 * Run suite

## Task descriptions

> Though derived from the standard Regional Nested Suite (see e.g. [Regional Practical 2023](https://code.metoffice.gov.uk/trac/jumps/wiki/RegionaPractical2023) (MOSRS account required)), the AUS2200 suite contains a number of changes designed to optimise the suite workflow. The model runs, whether continuous or restarted, are divided into 6 hour cycles, with the task graph for each cycle represented in the following image

Author: The setup step still needs to be filled out. Those are just dot points to remind me what to write

> Though derived from the standard Regional Nested Suite (see e.g. [Regional Practical 2023](https://code.metoffice.gov.uk/trac/jumps/wiki/RegionaPractical2023) (MOSRS account required)), the AUS2200 suite contains a number of changes designed to optimise the suite workflow. The model runs, whether continuous or restarted, are divided into 6 hour cycles, with the task graph for each cycle represented in the following image

Reviewer: I got that, the comment was more about remove MOSRS account .. from here

By default these tasks are configured to run on 2 Sapphire Rapids nodes (`normalsr` queue, 208 cpus) each and complete in approximately 38 minutes. The tasks are independent, and are configured to run in parallel. This block of tasks is the second most expensive task set of the workflow, at a cost of about 1.8kSU/cycle. The output is saved to `{{CYLC_SUITE_SHARE_DIR}}/cycle/{{CYLC_TASK_CYCLE_POINT}}/ec/um`, and can be cleared after the forecast task has been completed.

### `aus2200_d0198_RA3_um_createbc_000`
This task takes the driving model reconfiguration output and creates the boundary 'frames' that are read by the model every hour to synchronise along the spatial boundaries. These frames are fairly small and data-derived, therefore are ideal for re-use in the 'restart' run mode where spin-up times and model runtimes overlap. This task is not run if the `get_lbc_from_archive` task in the same cycle was successful. The task takes approximately 5 minutes to complete on a single core.

### `put_lbc_to_archive`

> This task archives the boundary conditions created by `aus2200_d0198_RA3_um_createbc_000` to the directory determined above in the `get_lbc_from_archive` task. Note that there is a race condition between testing whether archived boundary conditions are available, to this task running. When running overlapping restart suites, it is best to ensure that the same cycle point is not running in two different suites to avoid this race condition. This task is not run if `Archive LBCS` is set to false in the 'Driving model setup' section of the rose suite configuration (`{{ARCHIVE_LBCS}}` in `rose-suite.conf`)

Reviewer: I find this part of the sentence confusing: "to the directory determined above in the `get_lbc_from_archive` task" as there's no reference to a directory being some kind of output in the get_lbc_from_archive section. Also maybe use "previously" instead of "above". You can probably cut it out as it's kind of obvious that the boundary conditions will have to be store somewhere, or mention the directory in the get_lbc_.. if this is an important detail

Reviewer: Also add a note somewhere of what a "race condition" is (even just a link to wikipedia?) as it won't be obvious for many researchers. And again I would put "This task is not run ..." before the Note on the race condition

### `recover_sst`

> This task retrieves the sea-surface temperature from input dump file created by the `ec_recon_000` task. This is necessary as there is an error in the sea ice concentration in the dump file caused by the interpolation from the low-resolution ERA5 dataset, combined with the `era5grib` library's handling of missing values. This can be summarised visually in the following images (Right image created by Martin Dix)

Reviewer: ```suggestion
This task retrieves the sea-surface temperature from input dump file created by the `ec_recon_000` task. This is necessary as there is an error in the sea ice concentration in the dump file caused by the interpolation with the `era5grib` library. This can be summarised visually in the following images (Right image created by Martin Dix)
```

Reviewer: My feeling is that it's too much detail better to create and link an issue in era5grib which hopefully eventually will be solved (if it's not due to xesmf) as opposed to put too much detail in here.

![suspicious sea ice](../Images/AUS2200/suspicious-sea-ice.png)

Any of the large resolution pixels that the reconfiguration step has determined to contain more than some fraction of ocean is treated as an ocean point with a sea ice fraction of about 14%. This error is rectified by restoring the sea ice fraction from the ancillary files, however the UM cannot do this without also loading the sea surface temperature from an ancillary file. This step ensures the sea surface temperature is not modified. The addition of these ancillary files into the input dump is controlled by the `2` and `3` entries under 'Configure ancils and initialise dump fields' section of the rose suite configuration
![recover sst](../Images/AUS2200/recover-sst.png)

### `create_soil_moisture_ancil`

> This step is used to carry over soil moisture from a previous run of the suite. When performing restart runs, it is advisable to carry the soil moisture through the entire run, as soil moisture has a much longer spin-up time than atmospheric processes. The restart run procedure is explained in more detail in the [Restart and Spinup](#restart-and-spinup) section below. This task is used to partially serialise restart runs, it can be submitted at any time and will wait until the correct dump file has been created from a previous run of the suite before creating the soil moisture ancillary file. The addition of this ancillary file into the input dump is controlled by the `1` entry under 'Configure ancils and initialise dump fields'.

Reviewer: The more I read the more I feel step is the right word for this as task should be a single action rather than a combination of. In any case step/task should be used consistently

> This step is used to carry over soil moisture from a previous run of the suite. When performing restart runs, it is advisable to carry the soil moisture through the entire run, as soil moisture has a much longer spin-up time than atmospheric processes. The restart run procedure is explained in more detail in the [Restart and Spinup](#restart-and-spinup) section below. This task is used to partially serialise restart runs, it can be submitted at any time and will wait until the correct dump file has been created from a previous run of the suite before creating the soil moisture ancillary file. The addition of this ancillary file into the input dump is controlled by the `1` entry under 'Configure ancils and initialise dump fields'.

Reviewer: ```suggestion
This step is used to carry over soil moisture from a previous run of the suite when performing restart runs. It is advisable to carry the soil moisture through the entire run, as soil moisture has a much longer spin-up time than atmospheric processes. The restart run procedure is explained in more detail in the [Restart and Spinup](#restart-and-spinup) section below. This task is used to partially serialise restart runs, it can be submitted at any time and will wait until the correct dump file has been created from a previous run of the suite before creating the soil moisture ancillary file. The addition of this ancillary file into the input dump is controlled by the `1` entry under 'Configure ancils and initialise dump fields'.
```

![soil moisture ancil](../Images/AUS2200/spinup-soil-mositure.png)

>  The creation of this ancillary file is controlled by the `SOIL_MOISTURE_CARRYOVER` setting under 'Cycling options' under the 'Nesting Suite' area in the rose suite configuration. Note that if this is set to false, the addition of the soil moisture field must be removed from the input dump. This is done by right-clicking on the row in the 'Configure ancils and initialise dump fields' area and selecting 'Ignore this section'.

Reviewer: ```suggestion
 The creation of this ancillary file is controlled by the `SOIL_MOISTURE_CARRYOVER` setting under 'Cycling options' under the 'Nesting Suite' area in the rose suite configuration. 
 Note that if this is set to false, the addition of the soil moisture field must be removed from the input dump. This is done by right-clicking on the row in the 'Configure ancils and initialise dump fields' area and selecting 'Ignore this section'.
```

The model runs in a 68x68 processor decomposition, with 2 threads per MPI process, 48 IO server tasks arranged in a 6x8 configuration (6 IO servers, 8 tasks per server). This gives a total of 4672 MPI processes, 9344 threads. In order to align with the 8 NUMA domains on the sapphire rapids nodes, the model is run with 96 threads per node, with 12 threads running on each NUMA domain. This gives a total of 98 nodes, with the final node utilised at one third capacity. 

A significant amount of optimisation work was performed in order to reach this level of performance. We advise not to change any of the UM processor decomposition or IO System Settings unless the current settings are preventing the model from running entirely.

### `housekeep_cycle`

This is a standard task in all UM suites that archives logs and cleans working directories. In this suite it is configured to tar the log files and remove the work directory from the previous cycle.

### `convert_to_nc`

> This task converts the raw UM output from the previous cycle to netCDF format. It is advised to retain this step, as the converted netCDF output tales up around 1/6th the disk space of the raw UM output. This step can also act as a 'copy output to final location' step. This task is only run when 'Convert output to netCDF' is set to true under 'General Run Options' (`{{CONVERT_TO_NETCDF}}` in `rose-suite.conf`). The final location of the netCDF files is set in the 'Path to converted netCDF files' configuration option (`{{NC_OUTPUT_PATH}}` in `rose-suite.conf`)

Reviewer: ```suggestion
This task converts the raw UM output from the previous cycle to netCDF format. It is advised to retain this step, as the converted netCDF output takes up around 1/6th the disk space of the raw UM output. This step removes the need to run the ACCESS-Archiver and can be used to move the output to its final location. This task is only run when 'Convert output to netCDF' is set to true under 'General Run Options' (`{{CONVERT_TO_NETCDF}}` in `rose-suite.conf`). The final location of the netCDF files is set in the 'Path to converted netCDF files' configuration option (`{{NC_OUTPUT_PATH}}` in `rose-suite.conf`)
```

Restart style runs are more complex, as each restart may require spin-up time, and output fields may need to be carried over between overlapping suite runs.The AUS2200 suite has the ability to manage this, specifically for soil moisture fields. The suite can be configured to carry over soil moisture data between 24-hour restart runs. This is controlled by a combination of the `create_soil_moisture_ancil` task, and the input into the `aus2200_d0198_RA3_um_recon` task. To enable soil moisture carry-over, set `SOIL_MOISTURE_CARRYOVER` to true, and set `SPINUP_HOURS` to an integer values 
![soil moisture carryover](../Images/AUS2200/soil-moisture-carryover.PNG).

Spin up hours is only used in restart runs, and determines how many hours of the model run are estimated to be required to 'spin-up' the atmosphere model. Typically, the model is restarted every 24 hours, and run for `24 + SPINUP_HOURS` model hours,  with the first `SPINUP_HOURS` discarded. In order for the soil moisture carryover to work, the suites must adhere to a strict naming convention that allows the `create_soil_moisture_ancil` task to derive the name of the previous restart suite. The convention is as follows:

> ```{note}

Reviewer: ```suggestion
```{admonition} Naming convention
```

Reviewer: This will create a box like for the note but with a title, I think something like this, feel free to use a better title will work better for this case, as you're referring to it beforehand it's more an example/clarification than a "note"

In your `~/roses` directory, make a copy of the AUS2200 suite, following the naming convention outlined [above](#restart-and-spinup). This example uses `u-cs142` for the original suite id.
<terminal-window lineDelay=0>
    <terminal-line data="output"><span style="color:#A6CE39">u-cs142</span><span style="color:#FAA619">\$&nbsp;</span>cd ..</terminal-line>
    <terminal-line data="output"><span style="color:#A6CE39">roses</span><span style="color:#FAA619">\$&nbsp;</span>cp -r u-cs142 u-cs142-20160601T0000</terminal-line>
    <terminal-line data="output"><span style="color:#A6CE39">roses</span><span style="color:#FAA619">\$&nbsp;</span>cd u-cs142-20160601T0000</terminal-line>
    <terminal-line data="output"><span style="color:#A6CE39">u-cs142-20160601T0000</span><span style="color:#FAA619">\$&nbsp;</span>rose edit</terminal-line>
</terminal-window>

> With the rose config editor window open, change `INITIAL_CYCLE_POINT` to match the date stamp of the directory, set `FINAL_CYCLE_POINT` to 20160602T1800. Re-enable `SOIL_MOISTURE_CARRYOVER` and ensure `SPINUP_HOURS` is set to 24. Ensure that item `1` in the 'Configure ancils and initialise dump fields' section of the `um` task has been re-enabled. 

Reviewer: should um be capitalised?

Author: No, this is the name of the section in the rose config editor, 'um' is lower case.

> ![suite configuration](../Images/AUS2200/suite-config.gif)

Reviewer: I'm just being fussy here but maybe slow this down a notch if possible?

Author: Can do, but will need to recreate the gif. The actual footage is 30 seconds, I'll try 50% speedup and see how that looks.

Reviewer: Not critical, wait and see what others think
